examples/alife-boid/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+mod boid;
+mod rand;
+mod triangle;
+mod vector;
+
+use crate::boid::Boid;
+use crate::rand::Rand;
+use crate::triangle::Triangle;
+use std::time::Duration;
+use yew::prelude::{html, Component, ComponentLink, Html, ShouldRender};
+use yew::services::{IntervalService, Task};
+
+pub struct Model {
+    boids: Vec<Boid>,
+    #[allow(unused)]
+    job: Box<dyn Task>,
+}
+
+pub enum Msg {
+    Tick,
+}
+
+const WIDTH: u64 = 600;
+const HEIGHT: u64 = 400;
+
+impl Component for Model {
+    type Message = Msg;
+    type Properties = ();
+
+    fn create(_: Self::Properties, link: ComponentLink<Self>) -> Self {
+        let callback = link.callback(|_| Msg::Tick);
+        let handle = IntervalService::spawn(Duration::from_millis(50), callback);
+
+        let mut rng = Rand::new();
+        Self {
+            boids: (0..100).map(|_| Boid::new(&mut rng)).collect(),
+            job: Box::new(handle),
+        }
+    }
+
+    fn update(&mut self, msg: Self::Message) -> ShouldRender {
+        match msg {
+            Msg::Tick => {
+                let boids = self.boids.clone();
+                for boid in &mut self.boids {
+                    boid.next_state(&boids);
+                }
```

Comment:
We can avoid cloning the vector by separating the velocity calculation from the position calculation.
The only step that still needs access to the boids slice after the optimisations I mentioned is `calc_separation` which only depends on the position, not the velocity. We can thus safely calculate the new velocity while mutating the prior boids and then apply the velocities in a separate step.

```suggestion
                let boids = &mut self.boids;
                // calculate the new velocity for each boid
                for boid in &mut boids {
                    // VELOCITY_SUM and POSITION_SUM are placeholders for the pre-calculated values for `calc_cohesion` and `calc_alignment`.
                    boid.update_velocity(&boids, VELOCITY_SUM, POSITION_SUM);
                }
 
                // apply the velocity for each boid
                for boid in &mut boids {
                    boid.move();
                }
```

Reply:
Sorry. I don't understand how I can pre-calculation. Is it a kind of inline function?

Reply:
Hmm, I've changed my mind on this one. I don't think it's worth the effort for this example.
Let's focus on the other things instead!